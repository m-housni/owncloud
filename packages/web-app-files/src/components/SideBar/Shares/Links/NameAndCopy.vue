<template>
  <div class="oc-mb-s oc-width-1-1">
    <h4 class="oc-text-truncate oc-text-normal oc-files-file-link-name oc-m-rm" v-text="linkName" />
    <div class="oc-flex oc-flex-middle oc-flex-between oc-width-1-1 oc-p-s link-name-container">
      <div v-if="copied" class="oc-flex oc-flex-middle oc-text-truncate">
        <oc-icon variation="success" name="checkbox-circle" />
        <p class="oc-files-file-link-url url-copied oc-ml-s oc-my-rm" v-text="copiedLabel" />
      </div>
      <div v-else class="oc-flex oc-flex-middle oc-text-truncate">
        <oc-icon name="link" fill-type="line" />
        <p
          v-oc-tooltip="link.url"
          class="oc-files-file-link-url oc-ml-s oc-text-truncate oc-my-rm"
          v-text="link.url"
        />
      </div>
      <oc-button
        v-oc-tooltip="copyBtnHint"
        appearance="raw"
        :aria-label="copyBtnHint"
        class="oc-files-public-link-copy-url"
        @click="copyLinkToClipboard"
        v-text="copyBtnLabel"
      />
    </div>
  </div>
</template>

<script>
import { mapActions } from 'vuex'
import { defineComponent } from 'vue'
import copyToClipboard from 'copy-to-clipboard'

export default defineComponent({
  name: 'NameAndCopy',
  props: {
    link: {
      type: Object,
      required: true
    }
  },
  data: () => ({
    copied: false,
    timeout: null
  }),
  computed: {
    linkName() {
      return this.link.name
    },
    copyBtnLabel() {
      return this.$gettext('Copy')
    },
    copyBtnHint() {
      return this.$gettext('Copy link to clipboard')
    },
    copiedLabel() {
      return this.$gettext('Copied')
    }
  },
  methods: {
    ...mapActions(['showMessage']),
    copyLinkToClipboard() {
      copyToClipboard(this.link.url)
      this.clipboardSuccessHandler()
      this.showMessage({
        title: this.link.quicklink
          ? this.$gettext('The quicklink has been copied to your clipboard.')
          : this.$gettextInterpolate(
              this.$gettext('The link "%{linkName}" has been copied to your clipboard.'),
              {
                linkName: this.linkName
              },
              true
            )
      })
    },
    clipboardSuccessHandler() {
      this.copied = true
      clearTimeout(this.timeout)
      this.timeout = setTimeout(() => {
        this.copied = false
      }, 550)
    }
  }
})
</script>

<style lang="scss" scoped>
.url-copied {
  color: var(--oc-color-swatch-success-default);
}
</style>
