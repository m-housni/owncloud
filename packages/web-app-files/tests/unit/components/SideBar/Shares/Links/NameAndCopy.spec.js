import { createLocalVue, mount } from '@vue/test-utils'
import Vuex from 'vuex'
import GetTextPlugin from 'vue-gettext'
import DesignSystem from '@ownclouders/design-system'
import NameAndCopy from 'web-app-files/src/components/SideBar/Shares/Links/NameAndCopy.vue'

jest.useFakeTimers()

const localVue = createLocalVue()

localVue.use(Vuex)
localVue.use(DesignSystem)
localVue.use(GetTextPlugin, {
  translations: 'does-not-matter.json',
  silent: true
})

const mapActions = {
  showMessage: jest.fn()
}

const exampleLink = {
  name: 'Example link',
  url: 'https://some-url.com/abc'
}

describe('NameAndCopy', () => {
  it('should show link info component including a copy-to-clipboard button', () => {
    const wrapper = getWrapper()
    expect(wrapper).toMatchSnapshot()
  })
  it('upon clicking it should copy the private link to the clipboard button, render a success message and change icon for half a second', async () => {
    const spyShowMessage = jest.spyOn(mapActions, 'showMessage')
    const windowSpy = jest.spyOn(window, 'prompt').mockImplementation()

    const store = createStore()
    const wrapper = getWrapper(store)

    expect(spyShowMessage).not.toHaveBeenCalled()
    expect(windowSpy).not.toHaveBeenCalled()

    await wrapper.find('.oc-files-public-link-copy-url').trigger('click')

    expect(wrapper).toMatchSnapshot()

    expect(spyShowMessage).toHaveBeenCalledTimes(1)
    expect(windowSpy).toHaveBeenCalledTimes(1)
    expect(windowSpy).toHaveBeenCalledWith('Copy to clipboard: Ctrl+C, Enter', exampleLink.url)

    jest.advanceTimersByTime(550)

    wrapper.vm.$nextTick(() => {
      expect(wrapper).toMatchSnapshot()
    })
  })
})

function createStore() {
  return new Vuex.Store({
    actions: mapActions,
    commit: jest.fn(),
    dispatch: jest.fn(),
    getters: {
      capabilities: function () {
        return {
          files: {
            privateLinks: true
          }
        }
      }
    }
  })
}

function getWrapper(store) {
  return mount(NameAndCopy, {
    localVue,
    store,
    propsData: {
      link: exampleLink
    },
    directives: {
      'oc-tooltip': jest.fn()
    }
  })
}
