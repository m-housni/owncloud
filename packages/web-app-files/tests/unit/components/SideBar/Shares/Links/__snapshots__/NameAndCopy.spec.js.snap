// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NameAndCopy should show link info component including a copy-to-clipboard button 1`] = `
<div class="oc-mb-s oc-width-1-1">
  <h4 class="oc-text-truncate oc-text-normal oc-files-file-link-name oc-m-rm">Example link</h4>
  <div class="oc-flex oc-flex-middle oc-flex-between oc-width-1-1 oc-p-s link-name-container">
    <div class="oc-flex oc-flex-middle oc-text-truncate"><span class="oc-icon oc-icon-m oc-icon-passive"><!----></span>
      <p class="oc-files-file-link-url oc-ml-s oc-text-truncate oc-my-rm">https://some-url.com/abc</p>
    </div> <button aria-label="Copy link to clipboard" type="button" class="oc-files-public-link-copy-url oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw">Copy</button>
  </div>
</div>
`;

exports[`NameAndCopy upon clicking it should copy the private link to the clipboard button, render a success message and change icon for half a second 1`] = `
<div class="oc-mb-s oc-width-1-1">
  <h4 class="oc-text-truncate oc-text-normal oc-files-file-link-name oc-m-rm">Example link</h4>
  <div class="oc-flex oc-flex-middle oc-flex-between oc-width-1-1 oc-p-s link-name-container">
    <div class="oc-flex oc-flex-middle oc-text-truncate"><span class="oc-icon oc-icon-m oc-icon-success"><!----></span>
      <p class="oc-files-file-link-url url-copied oc-ml-s oc-my-rm">Copied</p>
    </div> <button aria-label="Copy link to clipboard" type="button" class="oc-files-public-link-copy-url oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw">Copy</button>
  </div>
</div>
`;

exports[`NameAndCopy upon clicking it should copy the private link to the clipboard button, render a success message and change icon for half a second 2`] = `
<div class="oc-mb-s oc-width-1-1">
  <h4 class="oc-text-truncate oc-text-normal oc-files-file-link-name oc-m-rm">Example link</h4>
  <div class="oc-flex oc-flex-middle oc-flex-between oc-width-1-1 oc-p-s link-name-container">
    <div class="oc-flex oc-flex-middle oc-text-truncate"><span class="oc-icon oc-icon-m oc-icon-passive"><!----></span>
      <p class="oc-files-file-link-url oc-ml-s oc-text-truncate oc-my-rm">https://some-url.com/abc</p>
    </div> <button aria-label="Copy link to clipboard" type="button" class="oc-files-public-link-copy-url oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw">Copy</button>
  </div>
</div>
`;
